### DISCUSSION TITLE
How to pass PrismaClient object to custom library to dynamically create prismaservice instance and create middleware

### DISCUSSION THREAD

**Discussion by hemendpatel** [prisma/prisma#11519] [Q&A]
## Problem

I’m working on micro service design and want to create a common logging and tracing middleware for prisma service call.
I want to create nestjs library to share the logging and tracing middleware, but library doesn’t have the PrismaClient instance.
Does anyone used Prisma to create library to dynamically inject configuration to create prismaservice instance and log query request and response data.

I.e
**PrismaModule**

```
@Module({ imports: [ConfigDataModule] })
export class PrismaModule {
    static register(client: any): DynamicModule {
        console.log('client', client)
        return {
            module: PrismaModule,
            providers: [
                {
                    provide: Client,
                    useClass: client
                }, PrismaService],
            exports: [PrismaService]

        }

    }
}
```

**PrismaService** 
```
@Injectable()
export class PrismaService extends Client implements OnModuleInit, OnModuleDestroy {
    constructor(private configDataService: ConfigDataService) {
        super({
            datasources: {
                db: {
                    url: `postgresql://${configDataService.postgresSettings}`,
                },
            },
            log: [
                { emit: 'event', level: 'query' }
                { emit: 'stdout', level: 'error' },
            ],
            errorFormat: 'colorless',
        });

        this.$on<any>('query', (event: QueryEvent) => {
            console.log('Query: ' + event.query);
            console.log('Params: ' + event.params);
            console.log('Duration: ' + event.duration + 'ms');
            console.log('Timestamp: ' + event.timestamp + 'ms');
        });
        this.$use(async (params: MiddlewareParams, next) => {
            const before = Date.now();
            const result = await next(params);
            const after = Date.now();
            console.log(`Query ${params.model}.${params.action} took ${after - before}ms`);
            console.log(`Result ${JSON.stringify(result)}`);
            return result;
        });
    }

    async onModuleInit() {
        await this.$connect();
    }

    async onModuleDestroy() {
        await this.$disconnect();
    }
}
```

**Comment by TasinIshmam** [comment]
Hi @hemendpatel 👋  

I'm not perfectly sure exactly what you want to do. Could you kindly elaborate a bit more? 

Your existing PrismaService/PrismaModule implementation is already logging all queries that are being made through Prisma. 

Do you want to create a package from this implementation that you can share across different microservices? 

Could you also elaborate on what you meant by "dynamically inject configuration"? My understanding is, you want to inject the configuration to instantiate Prisma Client. Is this correct?  

If I'm mistaken about anything, please let me know.

**Reply by vladi-stevanovic** [reply]
:wave: Hello @hemendpatel , I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
